lib/Transform/Clang/RemoveFirstPrivate.cpp

```diff
@@ -0,0 +1,349 @@
+        }
+        varStack.top().rvalName = varName;
+        if (varStack.top().dimensionsNum > 0) {
+          waitingForDimensions = true;
+        }
+
+      }
+      waitingForVar = !waitingForVar;
+    }
+    return RecursiveASTVisitor::TraverseDeclRefExpr(Ex);
+  }
+
+  bool TraverseIntegerLiteral(IntegerLiteral *IL) {
+
+    if (isInPragma) {
+      if (waitingForDimensions && curDimensionNum == varStack.top().dimensionsNum) {
+        waitingForDimensions = false;
+        curDimensionNum = 0;
+      }
+      int val = IL -> getValue().getLimitedValue();
+      if (waitingForDimensions) {
+        if (varStack.size()) {
+          varStack.top().dimensions.push_back(val);
+          curDimensionNum++;
+        }
+      } else if (!waitingForVar) {    // get rvalue
+        varStack.top().rvalName = std::to_string(val);
+        waitingForVar = !waitingForVar;
+        if (varStack.top().dimensionsNum > 0) {
+          waitingForDimensions = true;
+        }
+      }
+    }
+    return RecursiveASTVisitor::TraverseIntegerLiteral(IL);
+  }
+
+#ifdef LLVM_DEBUG
+// debug info
+#endif
+
+private:
+  /// Return current scope.
+  Stmt *getScope() {
+    for (auto I = mScopes.rbegin(), EI = mScopes.rend(); I != EI; ++I)
+      if (isa<ForStmt>(*I) || isa<CompoundStmt>(*I))
+        return *I;
+    return nullptr;
+  }
+
+  bool isInPragma = false;
+  bool waitingForVar = true;
+  bool waitingForDimensions = false;
+  int curDimensionNum = 0;
+
+  std::vector<Stmt*> mScopes;
+
+  TransformationContext *mTfmCtx;
+  const ASTImportInfo &mImportInfo;
+  ClangGlobalInfoPass::RawInfo *mRawInfo;
+  Rewriter &mRewriter;
+  ASTContext &mContext;
+  SourceManager &mSrcMgr;
+  const LangOptions &mLangOpts;
+  SmallVector<Stmt *, 1> mClauses;
+
+  std::stack<vars> varStack;
+
+};
+}
+
+bool ClangRemoveFirstPrivate::runOnFunction(Function &F) {
+  auto *M = F.getParent();
+
+  auto *DISub{findMetadata(&F)};
+  if (!DISub)
+    return false;
+  auto *CU{DISub->getUnit()};
+  if (isC(CU->getSourceLanguage()) && isCXX(CU->getSourceLanguage()))
+    return false;
+  auto &TfmInfo{getAnalysis<TransformationEnginePass>()};
+  auto *TfmCtx{TfmInfo ? dyn_cast_or_null<ClangTransformationContext>(
+                             TfmInfo->getContext(*CU))
+                       : nullptr};
+  if (!TfmCtx || !TfmCtx->hasInstance()) {
+    F.getContext().emitError(
+        "cannot transform sources"
+        ": transformation context is not available for the '" +
+        F.getName() + "' function");
+    return false;
+  }
+
+  auto FuncDecl = TfmCtx->getDeclForMangledName(F.getName());
+  if (!FuncDecl)
+    return false;
+
+  ASTImportInfo ImportStub;
+  const auto *ImportInfo = &ImportStub;
+  if (auto *ImportPass = getAnalysisIfAvailable<ImmutableASTImportInfoPass>())
+    ImportInfo = &ImportPass->getImportInfo();
+  auto &GIP = getAnalysis<ClangGlobalInfoPass>();
+
+  DeclVisitor Visitor(*TfmCtx, *ImportInfo, GIP.getRawInfo());
```

Comment:
После добваления поддержки многофайловых проектов интерфейс `GlobalInfoPass` усложнился.
```suggestion
  auto *GI{GIP.getGlobalInfo(TfmCtx)};
  assert(GI && "Global declaration must be collected!");
  DeclVisitor Visitor(*TfmCtx, *ImportInfo, GI->RI);
```